### PR TITLE
Add GitOps application details page

### PR DESCRIPTION
**Pull request by divyanshiGupta**
Jira Story - https://issues.redhat.com/browse/ODC-4289

This PR adds GitOps application details page

Screenshot -
image

Setup for testing the PR -

- Install the GitOps operator from https://github.com/redhat-developer/gitops-operator
  Note: Use OCP >= v4.5
- Fetch the proxy PR 
 `git fetch upstream refs/pull/5928/head && git checkout -b gitops  FETCH_HEAD`
- Build backend to enable the proxy
   ```
   source ./contrib/oc-environment.sh
   ./build-backend.sh 
   ./bin/bridge
   ```
- Create a namespace 
`kubectl create namespace pipelines-{USERNAME}-github`

- Create secret with your github token in above namespace
`kubectl create secret generic SECRET_NAME --from-literal=token=${GITHUB_TOKEN} -n SECRET_NS`

- Add the below mentioned annotation in your namespace
`app.openshift.io/vcs-uri: https://github.com/rhd-gitops-example/gitops.git`

Blockers

- GitOps proxy PR - https://github.com/openshift/console/pull/5928
- https://github.com/redhat-developer/gitops-operator/pull/3

### CODE REVIEW

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.scss [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    font-size: 18px;
+  }
+  &__app-pod {
+    width: 80px;
+    font-size: 12px;
+    padding: 10px 13px 10px 0px;
+  }
+  &__app-pr-details {
+    font-size: 12px;
+    padding-top: 20px;
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.scss:13]: please use Pf variables wherever possible?

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import './GitOpsServiceDetailsSection.scss';
+
+interface GitOpsServiceDetailsSectionProps {
+  services: any;
+}
+
+const GitOpsServiceDetailsSection: React.FC<GitOpsServiceDetailsSectionProps> = ({ services }) => {
+  return (
+    <>
+      {_.map(services, (service) => (
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx:23]: no need to use fragment you can directly return _.map

File: frontend/packages/dev-console/src/components/gitops/GitOpsDetailsPage.tsx [hunk 3]
```diff
@@ -1,22 +1,126 @@
...
+            <span className="co-resource-item__resource-name">{appName}</span>
+          </div>
+        </h1>
+        <Split style={{ alignItems: 'center' }} hasGutter>
+          <SplitItem style={{ fontWeight: 'bold', fontSize: '12px' }}>
+            Manifest File Repo:
+          </SplitItem>
+          <SplitItem isFilled>
+            <Label
+              style={{ fontSize: '12px' }}
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/GitOpsDetailsPage.tsx:112]: move static styles to scss

File: frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import * as React from 'react';
+import { PodModel } from '@console/internal/models';
+import { useK8sWatchResource } from '@console/internal/components/utils/k8s-watch-hook';
+import { K8sResourceKind, modelFor } from '@console/internal/module/k8s';
+import { PodRing, ExtPodKind } from '@console/shared';
+import { LoadingInline } from '@console/internal/components/utils';
+import * as _ from 'lodash';
+
+interface PodRingWrapperProps {
+  workload: { [key: string]: any };
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx:10]: need proper type here

File: frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+  const podResource = React.useMemo(
+    () => ({
+      kind: PodModel.kind,
+      namespace: workload?.namespace,
+      selector: workloadData?.spec?.selector,
+      isList: true,
+      optional: true,
+    }),
+    [workload, workloadData],
+  );
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx:35]: add checks for data is loaded or not
```suggestion
  const [workloadData, loaded, loadError] = useK8sWatchResource<K8sResourceKind>(workloadResource);

  const podResource = React.useMemo(
    () => (loaded && !loadError ? {
      kind: PodModel.kind,
      namespace: workload?.namespace,
      selector: workloadData?.spec?.selector,
      isList: true,
      optional: true,
    }: {}),
    [workload, workloadData, loaded, loadError],
  );
```

File: frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+import { useK8sWatchResources } from '@console/internal/components/utils/k8s-watch-hook';
+import { Timestamp } from '@console/internal/components/utils';
+import { K8sResourceKind } from '@console/internal/module/k8s';
+
+interface TimestampWrapperProps {
+  resModels: any;
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx:8]: need proper type

File: frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+const TimestampWrapper: React.FC<TimestampWrapperProps> = ({ resModels }) => {
+  const [lastUpdatedTimestamp, setLastUpdatedTimestamp] = React.useState(null);
+  const memoResources = React.useMemo(() => {
+    let resources = {};
+    _.forEach(resModels, (res) => {
+      const { group, version, kind, name, namespace } = res;
+      resources = {
+        ...resources,
+        [`${name}-${kind}-${namespace}`]: {
+          kind: [group, version, kind].join('~'),
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx:20]:
```suggestion
          kind: referenceForModel(res),
```

File: frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        return new Date(resTimestamp);
+      }),
+    );
+    setLastUpdatedTimestamp(timestamp);
+  }, [resourcesData]);
+
+  return (
+    <>
+      {lastUpdatedTimestamp ? (
+        <Timestamp timestamp={lastUpdatedTimestamp} />
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/TimestampWrapper.tsx:47]: not need to use Fragment here

File: frontend/packages/dev-console/src/components/gitops/GitOpsDetailsPage.tsx [hunk 9]
```diff
@@ -1,22 +1,126 @@
...
+  const searchParams = new URLSearchParams(location.search);
+  const manifestURL = searchParams.get('url');
+  const userName = _.replace(user?.metadata?.name ?? '', /[^a-zA-Z0-9-]/g, '');
+  const [secretNS, secretName] =
+    (userName && [`pipelines-${userName}-github`, `${userName}-github-token`]) || [];
+  const pipelinesBaseURL =
+    (secretNS &&
+      secretName &&
+      `/api/gitops/pipelines?secretNS=${secretNS}&secretName=${secretName}`) ||
+    undefined;
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/dev-console/src/components/gitops/GitOpsDetailsPage.tsx:35]: I'm fine with this but just a suggestion, how about creating utility for base URL and URIs

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        <StackItem className="odc-gitops-service-details">
+          <Card>
+            <CardTitle className="odc-gitops-service-details__app-title">
+              {(service?.workloadKind && (
+                <ResourceLink kind={service?.workloadKind} name={service?.name} linkTo={false} />
+              )) ?? <span className="co-resource-item__resource-name">{service?.name}</span>}
+            </CardTitle>
+            <CardBody className="odc-gitops-service-details__text">
+              <Label style={{ fontSize: '12px' }} color="cyan">
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx:32]: move static style to scss

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+const GitOpsServiceDetailsSection: React.FC<GitOpsServiceDetailsSectionProps> = ({ services }) => {
+  return (
+    <>
+      {_.map(services, (service) => (
+        <StackItem className="odc-gitops-service-details">
+          <Card>
+            <CardTitle className="odc-gitops-service-details__app-title">
+              {(service?.workloadKind && (
+                <ResourceLink kind={service?.workloadKind} name={service?.name} linkTo={false} />
+              )) ?? <span className="co-resource-item__resource-name">{service?.name}</span>}
```
Comment (sahil143) [inline frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx:29]:
```suggestion
              {(service?.workloadKind ? (
                <ResourceLink kind={service?.workloadKind} name={service?.name} linkTo={false} />
              )) : <span className="co-resource-item__resource-name">{service?.name}</span>}
```

**Comment by divyanshiGupta** [comment]
@sahil143 Thanks for the review. Seems like most of the comments are for code clean-up which I planned to after all the pieces were in place. Though most of the work is done now, so I am working on the clean-up now. There were some comments because I am interchangeably testing with mock and live data.

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx (hunk 2 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.tsx:23]: It doesnt seem like we can directly return map

File: frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.scss (hunk 1 above)
Comment (rohitkrai03) [reply frontend/packages/dev-console/src/components/gitops/GitOpsServiceDetailsSection.scss]: I don't see anything aligning to PF variables value here. The font size could be `--pf-global--FontSize--xs` but as you pointed out in your other PR, there's discrepancies in the docs and actual values.

File: frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx (hunk 5 above)
Comment (divyanshiGupta) [reply frontend/packages/dev-console/src/components/gitops/PodRingWrapper.tsx:35]: cannot add empty object here

**Comment by serenamarie125** [comment]
hey @divyanshiGupta this is looking really good!  It seems like there are some inconsistencies with spacing, etc.  Are you going to be able to address those before FF or will It be in a follow-on?

Glad to see you're able to get 5 env on a page without scrolling 🎉 
FYI @bkrikori

**Comment by divyanshiGupta** [comment]
@serenamarie125 I have added the missing spaces already, forgot to update the mock. Will update it now.

**Comment by divyanshiGupta** [comment]
@serenamarie125 updated.

**Comment by serenamarie125** [comment]
One more nit, are you able to the empty state before FF as well?
Screen Shot 2020-07-30 at 3 09 54 PM

Beth added a single message which runs the width of the card.

**Comment by divyanshiGupta** [comment]
@serenamarie125 I have mentioned in the doc that it is possible to have pod-details and not have commit-details or vice-versa. So I think it is better to have two separate empty states for both.

**Pull request merged**